src/devtools/client/debugger/src/components/Editor/Breakpoints/Panel/Panel.tsx

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
+
+import React, { useEffect, useState } from "react";
+import classnames from "classnames";
+import PanelEditor from "./PanelEditor";
+import BreakpointNavigation from "devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointNavigation";
+import Widget from "./Widget";
+import { useDispatch, useSelector } from "react-redux";
+import { inBreakpointPanel } from "devtools/client/debugger/src/utils/editor";
+import PanelSummary from "./PanelSummary";
+import FirstEditNag from "./FirstEditNag";
+import hooks from "ui/hooks";
+import { Nag } from "ui/hooks/users";
+import { prefs } from "ui/utils/prefs";
+import { Breakpoint, getHitCountsForSelectedSource } from "devtools/client/debugger/src/selectors";
+import { getAnalysisPointsForBreakpoint } from "ui/reducers/app";
+import { UIState } from "ui/state";
+import { clearHoveredItem, setHoveredItem } from "ui/actions/timeline";
+import { Editor } from "codemirror";
+import { AnalysisError, AnalysisPayload } from "ui/state/app";
+
+export type Input = "condition" | "logValue";
+type CodeMirror = any;
+
+function getPanelWidth(editor: Editor) {
+  // The indent value is an adjustment for the distance from the gutter's left edge
+  // to the panel's left edge, which is approximately ~60.
+  const panelIndent = 60;
+  return editor.getScrollInfo().clientWidth - panelIndent;
+}
+
+export enum PanelErrors {
+  TooManyPoints = "too-many-points",
+  MaxHits = "max-hits",
+  Default = "default",
+}
+
+const getPanelError = (hits: number | null, analysisPoints: AnalysisPayload | null) => {
+  if (analysisPoints?.error === AnalysisError.TooManyPoints) {
+    return PanelErrors.TooManyPoints;
+  } else if (analysisPoints?.error === AnalysisError.Default) {
+    return PanelErrors.Default;
+  } else if (hits && hits > prefs.maxHitsDisplayed) {
+    return PanelErrors.MaxHits;
+  }
+
+  return null;
+};
+
+const useGetHitsForPanel = (breakpoint: Breakpoint) => {
+  const analysisPoints = useSelector((state: UIState) =>
+    getAnalysisPointsForBreakpoint(state, breakpoint)
+  );
+  const hitCounts = useSelector(getHitCountsForSelectedSource);
+  const hitCountFromSources = hitCounts?.find(
+    hitCount => hitCount.location.line === breakpoint.location.line
+  );
+
+  const loading = analysisPoints === null || !hitCountFromSources;
+  const hits = analysisPoints?.data.length ?? hitCountFromSources?.hits ?? null;
```

Comment:
Ooo, this is interesting. Two sources of truth for the number of hits. I don't think I understand the surrounding code well enough to have an informed opinion on this but it jumps out to me.

Reply:
The summary here is that there are two ways of knowing how many times a particular location was hit. The first, older, and more thorough way is by running an analysis on that location. This will only return results for the given location, and it will also include tons of other information about the hits, not just the count. The second, newer, but sometimes more finicky route is using `heatMaps` (under the hood the `getHitCounts` [protocol message](https://static.replay.io/protocol/tot/Debugger/#method-getHitCounts). We switch between the two based on:

- Features the user has enabled (you can turn heatMaps off)
- How old the recording is (older recordings don't support heatMaps)
- How many locations the file will query for (files with thousands of locations can trigger an error on the backend when querying for heatMaps)

In all cases we fall back to running analyses.